src/toolbar/Toolbar.ts

```diff
@@ -0,0 +1,159 @@
+import { Dimensions } from '@dojo/widget-core/meta/Dimensions';
+import { DNode, WidgetProperties } from '@dojo/widget-core/interfaces';
+import { ThemedMixin, theme } from '@dojo/widget-core/mixins/Themed';
+import { v, w } from '@dojo/widget-core/d';
+import { WidgetBase } from '@dojo/widget-core/WidgetBase';
+
+import SlidePane, { Align } from '../slidepane/SlidePane';
+import * as css from './styles/toolbar.m.css';
+import * as iconCss from '../common/styles/icons.m.css';
+
+/**
+ * @type ToolbarProperties
+ *
+ * Properties that can be set on a Toolbar component
+ *
+ * @property actions           Action elements to show in the toolbar
+ * @property collapseWidth     Width at which to collapse actions into a SlidePane
+ * @property fixed             Fixes the toolbar to the top of the viewport
+ * @property onCollapse        Called when action items change their layout
+ * @property title             Element to show as the toolbar title
+ */
+export interface ToolbarProperties extends WidgetProperties {
+	actions?: DNode[];
+	collapseWidth?: number;
+	fixed?: boolean;
+	onCollapse?(collapsed: boolean): void;
+	title?: DNode;
+};
+
+export const ThemedBase = ThemedMixin(WidgetBase);
+
+@theme(css)
+@theme(iconCss)
+export default class Toolbar extends ThemedBase<ToolbarProperties> {
+	private _collapsed = false;
+	private _open = false;
+
+	constructor() {
+		super();
+		window.addEventListener('resize', this._collapseIfNecessary);
+	}
+
+	private _closeMenu() {
+		this._open = false;
+		this.invalidate();
+	}
+
+	private _collapseIfNecessary = () => {
+		const { collapseWidth = 800, onCollapse } = this.properties;
+		const { width } = this.meta(Dimensions).get('root').size;
+
+		if (width > collapseWidth && this._collapsed === true) {
+			this._collapsed = false;
+			onCollapse && onCollapse(this._collapsed);
+			this.invalidate();
+		}
+		else if (width <= collapseWidth && this._collapsed === false) {
+			this._collapsed = true;
+			onCollapse && onCollapse(this._collapsed);
+			this.invalidate();
+		}
+	}
+
+	private _toggleMenu() {
+		this._open = !this._open;
+		this.invalidate();
+	}
+
+	protected getFixedRootClasses(): (string | null)[] {
+		return [
+			css.rootFixed,
+			this.properties.fixed ? css.stickyFixed : null
+		];
+	}
+
+	protected getRootClasses(): (string | null)[] {
+		return [
+			css.root,
+			this._collapsed ? css.collapsed : null,
+			this.properties.fixed ? css.sticky : null
+		];
+	}
+
+	protected onAttach() {
+		this._collapseIfNecessary();
+	}
+
+	protected onDetach() {
+		window.removeEventListener('resize', this._collapseIfNecessary);
+	}
+
+	protected renderActions(): DNode {
+		const { actions = [], theme } = this.properties;
+
+		const actionsElements = actions.map((action, index) => v('div', {
+			classes: [ this.theme(css.action) ],
+			key: index
+		}, [ action ]));
```

Comment:
I don't really understand why these need to be wrapped in a div. Is it just so their css/theming is within the toolbar CSS as opposed to in parent-widget-level CSS?

Reply:
Yes, this is really just for easier theming. Most situations I thought of were pretty difficult to accomplish without actions being grouped in a common parent. I'm open to removing it if necessary though.

Reply:
Hm, I kind of imagine most of the time the actions will be passed in already wrapped in a parent, for semantics and styling. E.g.:
```
w(Toolbar, {
   actions: [
      v('div', { id: 'menu', role: 'menubar' }, [
         v('a', { href: '/#home' }, [ 'Home' ]),
         v('a', { href: '/#about' }, [ 'About' ]),
         v('a', { href: '/#contact' }, [ 'Contact' ])
      ])
   ],
   title: 'Foobar'
}, [
   content
])
```
In that case the extra wrapping div seems a little unnecessary. I don't really dislike it, I just think it introduces an extra wrapping element where none is needed.